### PR TITLE
add Feature/debugger detection

### PR DESCRIPTION
**Pull request by ha-ja**

### CODE REVIEW

File: Sources/ThreatDetectionCenter.swift [hunk 1]
```diff
@@ -14,6 +14,15 @@ public final class ThreatDetectionCenter {
         SimulatorDetection.threatDetected()
     }
     
+    /// Checks whether your application is being traced by a debugger.
+    ///
+    /// A debugger is a tool that allows developers to inspect and modify the execution of a program in real-time, potentially exposing sensitive data or allowing unauthorized control.
+    ///
+    /// Please note that Apple itself uses a debugger during a review and you could prevent Apple from carrying out a complete review. Which Apple will certainly not like.
```
Comment (theDeniZ) [inline Sources/ThreatDetectionCenter.swift:21]: our approach is not preventive: we only react to a debugger with no possible prevention methods implemented. I would remove or rephrase this sentence, maybe like "Note: Debugger may be required in Apple review process" or something like that

File: Sources/DebuggerDetection.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        hasTracerFlagSet()
+    }
+}
+
+// MARK: - Private
+fileprivate extension DebuggerDetection {
+    
+    /// Check P_TRACED flag from Darwin Kernel
+    /// if the process is traced
+    static func hasTracerFlagSet() -> Bool {
```
Comment (theDeniZ) [inline Sources/DebuggerDetection.swift:17]: it would be better to declare it as `private static`, i believe it will work, as the function is already in a fileprivate extensions

**Review by theDeniZ — CHANGES_REQUESTED** [review]

File: Sources/DebuggerDetection.swift [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        var info = kinfo_proc()
+        var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()] // Kernel info, process info, specific process by PID, get current process ID
+        var size = MemoryLayout.stride(ofValue: info)
+        
+        let result = sysctl(&mib, u_int(mib.count), &info, &size, nil, 0)
+        if result != 0 {
+            fatalError("sysctl failed")
```
[the remark this answers is not in the document]
Comment (ha-ja) [reply Sources/DebuggerDetection.swift:24]: Sure, I will change it to a false return at this point (no debugger recognized)?

Occurs if the sysctl call is not successful. Should not actually happen.

At this moment we cannot make a reliable statement about whether a debugger is running or not. It is not clear to me whether it would be better to return false or true. It is actually undefined at this point.

File: Sources/DebuggerDetection.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import Foundation
+import Darwin
```
[the remark this answers is not in the document]
Comment (ha-ja) [reply Sources/DebuggerDetection.swift:2]: We use a call of Darwin.sys.sysctl here:

https://github.com/EXXETA/iOS-Security-Toolkit/blob/48d5c6b438cdaeb2a55b9b3476652ae68eb62728/Sources/DebuggerDetection.swift#L18

https://github.com/EXXETA/iOS-Security-Toolkit/blob/48d5c6b438cdaeb2a55b9b3476652ae68eb62728/Sources/DebuggerDetection.swift#L22

However, this function is also available without explicitly specifying Import Darwin. 
Can be removed.

[the remark this answers is not in the document]
Comment (ha-ja) [reply Sources/DebuggerDetection.swift:2]: Should be removed?

File: Sources/DebuggerDetection.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ha-ja) [reply Sources/DebuggerDetection.swift:24]: return false at this point?

[the remark this answers is not in the document]
Comment (theDeniZ) [reply Sources/DebuggerDetection.swift:23]: yes, in this case we have only 2 options: assume no debugger is recognized (return false) or handle this case separately (ex. return nil). As the usecase should not be too often, it is fine to return here false.

[the remark this answers is not in the document]
Comment (jabou) [reply Sources/DebuggerDetection.swift:23]: Hi @theDeniZ ,

I'm checking this SDK as an option to implement it into our apps, but before we do that, I'm tracking this PR as we already have (Apple's) debugger check in the old SDK.

Usually, if the result is not 0, that means that we can not determine whether a debugger is connected. In that case, we can assume that checking the debugger's connection is not reliable. 
This leads me to be concerned about returning any value here.

If we look at Apple's implementation (https://developer.apple.com/library/archive/qa/qa1361/_index.html), they use assert to check that value. When the app is connected to the debugger (e.g., Xcode) and run from there, it's ok to have crash on that line, while on the production build (archive) this will be ignored.


TL;DR;

I would propose using the `assert(result != 0, "Sysctl failed, check could not be reliable")` instead of falsely returning `false`. 
Returning an option value would lead to more extensive documentation and a more complex call site, where SDK callers would have to handle all those use cases they could easily miss. Throwing an error is also an option, but it may be overkill for this case.

[the remark this answers is not in the document]
Comment (theDeniZ) [reply Sources/DebuggerDetection.swift:23]: Hi @jabou ,
thank you for sharing your thoughts and experience!
I understand the proposition and find it a very creative solution.

In a direct discussion with @ha-ja we talked about a need to correctly communicate the result of a "detection" methods, as not all of them are able to return a hard "yes"/"no" response. 
We would still like to offer the best api for every use-case, meaning, that everyone can be satisfied. 
I believe, deciding to intentionally crash the app is not a good choice (even considering differences in debug / release builds). If the consumer is comfortable with app crashes, it still is possible to assert the result.

We would like to offer 2 possible solutions and want to know your opinion, which would suit the best:
1. return nil, whenever no conclusive result can be found
2. return an Enum, describing the detection state

Both have pros and cons: nil is clean implementation which becomes ambiguous when multiple states result in nil; enums are a bit too much for a simple use-cases but communicate the state in the fullest (when properly documented)

I also ask @ha-ja  and @ychoura to vote as well. For convenience, use following reaction emoji for corresponding options:
1. 👍
2. 🎉

[the remark this answers is not in the document]
Comment (ychoura) [reply Sources/DebuggerDetection.swift:23]: I think crashing the app is not our decision to make. What I think we should do instead is return an optional bool and add docs. I find an enum to be a bit of an overkill at this stage of dev.

[the remark this answers is not in the document]
Comment (ha-ja) [reply Sources/DebuggerDetection.swift:23]: Thank you all for the discussion 👍

[the remark this answers is not in the document]
Comment (ha-ja) [reply Sources/DebuggerDetection.swift]: @alle
I have pushed the changes. 
isDebuggerDetected return is now an Optional Bool.

File: Sources/ThreatDetectionCenter.swift [hunk 5]
```diff
@@ -40,6 +56,10 @@ public final class ThreatDetectionCenter {
                 continuation.yield(.simulator)
             }
             
+            if DebuggerDetection.threatDetected() ?? false {
+                continuation.yield(.debugger)
+            }
```
Comment (ha-ja) [inline Sources/ThreatDetectionCenter.swift:61]: It is currently not possible to cover the nil return of DebuggerDetection.threatDetected() via the AsyncStream. 
If you want to check for nil, you must use the static function isDebuggerDetected without AsyncStream.

File: Sources/DebuggerDetection.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ha-ja) [reply Sources/DebuggerDetection.swift]: Please also note the following comment on AsyncStream.

https://github.com/EXXETA/iOS-Security-Toolkit/pull/10/files#r1932347230
> It is currently not possible to cover the nil return of DebuggerDetection.threatDetected() via the AsyncStream. If you want to check for nil, you must use the static function isDebuggerDetected without AsyncStream.

**Review by theDeniZ — APPROVED** [review]
thank you, approved
